src/libcollections/vec_map.rs

```diff
@@ -21,14 +21,44 @@ use core::fmt;
 use core::hash::{Hash, Writer, Hasher};
 use core::iter::{Enumerate, FilterMap, Map, FromIterator};
 use core::iter;
+use core::marker::InvariantType;
 use core::mem::replace;
+use core::mem;
 use core::ops::{Index, IndexMut};
 
-use {vec, slice};
-use vec::Vec;
+use {vec, slice, Vec};
 
 // FIXME(conventions): capacity management???
 
+/// This is the trait to bound on for keys of `VecMap`s. It cannot be
+/// implemented manually currently.
+pub trait VecMapKey {
+    #[doc(hidden)]
+    fn to_uint(&self) -> uint;
+    #[doc(hidden)]
+    fn from_uint(value: uint) -> Self;
+    #[doc(hidden)]
+    fn unimplementable(_: InvariantType<Self>) -> Unconstructible;
+}
+
+#[allow(missing_copy_implementations)]
+#[doc(hidden)]
+pub struct Unconstructible(());
+
+// Tell inference to not infer `uint`.
+struct _Nothing;
+impl VecMapKey for _Nothing {
```

Comment:
why not implement this for `u32` and other such types too instead of this dummy type?